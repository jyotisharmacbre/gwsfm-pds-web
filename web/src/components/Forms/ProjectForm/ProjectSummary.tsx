import React, { useState, useEffect } from 'react';
import { IProjectDetail } from '../../../store/CustomerEnquiryForm/Types/IProjectDetail';
import { ILookup } from '../../../store/Lookups/Types/ILookup';
import { LookupType } from '../../../store/Lookups/Types/LookupType';
import Company from '../../../store/DynamicsData/InitialState/Company';
import { toast } from 'react-toastify';
import {
	getPropertyName,
	getFilterElementFromArray,
	calculateRank,
	displayUserName
} from '../../../helpers/utility-helper';
import { FormattedMessage } from 'react-intl';
import { IUserServiceData } from '../../../store/UserService/Types/IUserService';
import * as services from '../../../services';
import { formatMessage } from '../../../Translations/connectedIntlProvider';
import { IDynamicsDivision, IDynamicBusinessUnits } from '../../../store/DynamicsData/Types/IDynamicData';

interface IProps {
	oneditclick: () => void;
	project: IProjectDetail;
	lookUpData: Array<ILookup>;
	currencySymbol: string;
	userNamesForEmails: Array<IUserServiceData>;
	handleGetUserNamesForEmails: (emails: Array<string>) => void;
	listOfDivisions: Array<IDynamicsDivision>;
	listOfBusinessUnits: Array<IDynamicBusinessUnits>;
	countryCode: string;
}
const ProjectSummary: React.FC<IProps> = (props) => {
	const [projectStatus, setProjectStatus] = useState<string>('');
	const [typeOfEngagement, setTypeOfEngagement] = useState<string>('');
	const companyObj = new Company();
	const [companyName, setCompanyName] = useState<string>('');
	const [headOfProject, setHeadOfProject] = useState<string>('');
	const [projectManager, setProjectManager] = useState<string>('');
	const [contractor, setContractor] = useState<string>('');
	const [division, setDivision] = useState<string>('');
	const [businessUnit, setBusinessUnit] = useState<string>('');

	useEffect(
		() => {
			let emails: Array<string> = [];
			if (props.project.projectId) {
				if (props.project.companyId == '0') setCompanyName(props.project.otherCompanyName);
				else
					services
						.getCompanies(props.project.companyId)
						.then((response) => {
							listOfCompaniesSuccess(response.data);
						})
						.catch((error) => {
							failure(error);
						});
				if (
					props.project.headOfProject &&
					emails.find((ele) => ele == props.project.headOfProject) == undefined
				)
					emails.push(props.project.headOfProject);
				if (
					props.project.projectManager &&
					emails.find((ele) => ele == props.project.projectManager) == undefined
				)
					emails.push(props.project.projectManager);
				props.handleGetUserNamesForEmails(emails);
				if (props.project.contractorId) {
					if (props.project.contractorId == '0') setContractor(props.project.otherContractName);
					else
						services
							.getContractsAndCustomers(props.project.contractorId)
							.then((response) => {
								getContractorSuccess(response.data);
							})
							.catch((error) => {
								failure(error);
							});
				}
				let filterDivision = props.listOfDivisions?.filter(
					(element) => element.divisionId == parseInt(props.project.divisionId)
				);
				if (filterDivision && filterDivision[0]) setDivision(filterDivision[0].description);
				let filterBusinessUnit = props.listOfBusinessUnits?.filter(
					(element) => element.businessUnitId == props.project.businessUnitId
				);
				if (filterBusinessUnit && filterBusinessUnit[0]) setBusinessUnit(filterBusinessUnit[0].description);
			}
		},
		[props.project, props.listOfDivisions, props.listOfBusinessUnits]
	);

	useEffect(
		() => {
			if (props.project.projectId && props.lookUpData && props.lookUpData.length > 0) {
				let filterStatus = props.lookUpData.filter(
					(element) =>
						element.lookupItem == LookupType.Project_Status && element.lookupKey == props.project.status
				);
				if (filterStatus && filterStatus[0]) setProjectStatus(filterStatus[0].description);
				let filterEngagementType = props.lookUpData.filter(
					(element) =>
						element.lookupItem == LookupType.Engagement_Type &&
						element.lookupKey == props.project.engagementId
				);
				let engagementLookupItem = filterEngagementType[0];
				if (filterEngagementType && engagementLookupItem) {
					const engagementVal = engagementLookupItem.lookupKey === 0 && props.project.otherEngagementType ? props.project.otherEngagementType : engagementLookupItem.description;
					setTypeOfEngagement(engagementVal);
				}
			}
		},
		[props.project, props.lookUpData]
	);

	useEffect(
		() => {
			if (props.userNamesForEmails && props.userNamesForEmails.length > 0) {
				let headOfProjectFilter = props.userNamesForEmails.find(
					(ele) => ele.email == props.project.headOfProject
				);
				if (headOfProjectFilter)
					setHeadOfProject(displayUserName(headOfProjectFilter));

				let projectManagerFilter = props.userNamesForEmails.find(
					(ele) => ele.email == props.project.projectManager
				);
				if (projectManagerFilter)
					setProjectManager(displayUserName(projectManagerFilter));
			}
		},
		[props.project, props.userNamesForEmails]
	);
	/* istanbul ignore next */
	const getContractorSuccess = (response) => {
		let filter = response.find((ele) => ele.contractId == props.project.contractorId);
		if (filter) setContractor(filter.contractName);
	};
	/* istanbul ignore next */
	const listOfCompaniesSuccess = (response) => {
		if (response && response.length > 0) {
			setCompanyName(
				getFilterElementFromArray(
					response,
					getPropertyName(companyObj, (prop) => prop.companyId),
					props.project.companyId,
					getPropertyName(companyObj, (prop) => prop.name)
				)
			);
		}
	};
	/* istanbul ignore next */
	const failure = error => {
		toast.error(formatMessage("MESSAGE_ERROR"));
	};


	return (
		<div className="RS_custom_block">
			<div className="title_edit_btn">
				<h4>
					<FormattedMessage id="HEADING_CUSTOMER_ENQUIRY" />
				</h4>
				<button type="submit" className="edit-btn" onClick={props.oneditclick} ><FormattedMessage id="BUTTON_EDIT" /></button>
			</div>

			<div className="RS_custom_inner">
				<div className="row">
					<div className="col-lg-4 col-sm-6">
						<ul>
							<li>
								<span>
									<FormattedMessage id="LABEL_CONTRACT" />
								</span>
								<p data-test="company">{contractor}</p>
							</li>
						</ul>
					</div>
					<div className="col-lg-4 col-sm-6">
						<ul>
							<li>
								<span>
									<FormattedMessage id="LABEL_DIVISION" />
								</span>
								<p data-test="company">{division}</p>
							</li>
						</ul>
					</div>
					<div className="col-lg-4 col-sm-6">
						<ul>
							<li>
								<span>
									<FormattedMessage id="LABEL_BUSINESS_UNIT" />
								</span>
								<p data-test="company">{businessUnit}</p>
							</li>
						</ul>
					</div>
					<div className="col-lg-4 col-sm-6">
						<ul>
							<li>
								<span>
									<FormattedMessage id="LABEL_PROJECT_MANAGER" />
								</span>
								<p>{projectManager}</p>
							</li>
						</ul>
					</div>
					<div className="col-lg-4 col-sm-6">
						<ul>
							<li>
								<span>
									<FormattedMessage id="LABEL_HEAD_OF_PROJECT" />
								</span>
								<p data-test="head-of-project">{headOfProject}</p>
							</li>
						</ul>
					</div>
					<div className="col-lg-4 col-sm-6">
						<ul>
							<li>
								<span>
									<FormattedMessage id="LABEL_CN_NUMBER" />
								</span>
								<p>{props.project.cnNumber}</p>
							</li>
						</ul>
					</div>
					<div className="col-lg-4 col-sm-6">
						<ul>
							<li>
								<span>
									<FormattedMessage id="LABEL_PROJECT" />
								</span>
								<p>{props.project.name}</p>
							</li>
						</ul>
					</div>
					{props.countryCode == "GBR" &&
						<div data-test="cdm_notifiable" className="col-lg-4 col-sm-6">
							<ul>
								<li>
									<span>
										<FormattedMessage id="LABEL_CDMNOTIFIABLE" />
									</span>
									<p>
										<FormattedMessage id={props.project.cdmNotifiable ? 'LABEL_YES' : 'LABEL_NO'} />
									</p>
								</li>
							</ul>
						</div>}
					<div className="col-lg-4 col-sm-6">
						<ul>
							<li>
								<span>
									<FormattedMessage id="LABEL_TYPE_OF_ENGAGEMENT" />
								</span>
								<p>{typeOfEngagement}</p>
							</li>
						</ul>
					</div>
					<div className="col-lg-4 col-sm-6">
						<ul>
							<li>
								<span>
									<FormattedMessage id="LABEL_PROBABILITY_OF_WINING" />
								</span>
								<p>{props.project.probabilityOfWinning}</p>
							</li>
						</ul>
					</div>
					<div className="col-lg-4 col-sm-6">	</div>
				</div>
				<div className="hr_line" />
				<div className="row">
					<div className="col-lg-12">
						<div className="scope_block">
							<h5>
								<FormattedMessage id="LABEL_PROJECT_SCOPE" />
							</h5>
							<ul>
								<li>{props.project.scope}</li>
							</ul>
						</div>
						<div className="comment_block">
							<h5>
								<FormattedMessage id="LABEL_COMMENTS" />
							</h5>
							<p>{props.project.comment}</p>
						</div>
					</div>
				</div>
			</div>
		</div>
	);
};

export default ProjectSummary;
