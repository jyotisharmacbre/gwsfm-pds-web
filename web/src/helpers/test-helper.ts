import checkPropTypes from 'check-prop-types';

export const findByTestAtrr = (component: any, attr: string) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};

export const checkProps = (component: any, expectedProps: any) => {
  const propsErr = checkPropTypes(
    component.propTypes,
    expectedProps,
    'props',
    component.name
  );
  return propsErr;
};
