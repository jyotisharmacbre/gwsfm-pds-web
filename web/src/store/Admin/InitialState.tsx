import { IAdminState } from './Types/IAdminState';
import { IAdminDefaults } from './Types/IAdminDefault';

export const initialState: IAdminState = {
	adminDefaultValues: [],
	error: ''
};
