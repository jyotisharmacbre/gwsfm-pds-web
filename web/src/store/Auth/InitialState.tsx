import { IAuth } from "./Types/IAuth";

export const initialState: IAuth= {
    token: null,
    name: null,
    firstName: null,
    email: null
};
