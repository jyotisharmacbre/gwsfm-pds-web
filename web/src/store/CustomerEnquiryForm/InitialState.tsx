import { IProjectDetailState } from './Types/IProjectDetailState';
import EventType from '../../enums/EventType';
import Notify from '../../enums/Notify';

export const initialState: IProjectDetailState = {
    form: {
        projectId: '',
        name: '',
        contractorId: '',
        companyId: '',
        headOfProject: '',
        projectOwner: '',
        projectManager: '',
        pmHasExperience: false,
        scope: '',
        cnNumber: '',
        status: 1,
        engagementId: undefined,
        countryId: 0,
        currencyId: 0,
        probabilityOfWinning: '',
        approxValue: '',
        contractTypeId: '',
        cdmNotifiable: false,
        firstAssetWorkedOn: undefined,
        secondAssetWorkedOn: undefined,
        thirdAssetWorkedOn: undefined,
        comment: '',
        otherCompanyName: '',
        otherContractName: '',
        divisionId: '',
        otherDivision: '',
        businessUnitId: '',
        otherBusinessUnit: '',
        weightedTCV: undefined,
        soldMargin: undefined,
        otherEngagementType: '',
        otherFirstAssetWorkedOn: '',
        otherSecondAssetWorkedOn: '',
        otherThirdAssetWorkedOn: ''
    },
    enquiryOverview: {
        projectName: '',
        contractorId: '',
        headOfProject: '',
        projectManager: '',
        scope: '',
        cnNumber: '',
        otherContractName: ''
    },
    error: null,
    loading: false,
    notify: Notify.none,
    event: EventType.none,
    enquiryOverviewError: null
};