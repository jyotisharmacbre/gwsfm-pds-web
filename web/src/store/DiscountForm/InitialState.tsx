import { IDiscountState } from './Types/IDiscountState';
import Notify from '../../enums/Notify';
import EventType from '../../enums/EventType';
import IDiscountSubContractor from './Types/IDiscountSubContractor';

export const newSubContractorDiscount: IDiscountSubContractor = {
  subContractorDiscountId: '',
  projectId: '',
  supplierName: '',
  supplierState: '',
  supplierTotalDiscount: undefined,
  supplierComments: '',
}

export const initialState: IDiscountState = {
  form: {
    projectId: '',
    clientDiscount: {
      discountId: '',
      projectId: '',
      discount: undefined,
      clientState: '',
      discountType: 1,
      clientComments: ''
    },
    subContractorDiscounts: [{ ...newSubContractorDiscount }],
  },
  error: null,
  loading: false,
  notify: Notify.none,
  event: EventType.none
};