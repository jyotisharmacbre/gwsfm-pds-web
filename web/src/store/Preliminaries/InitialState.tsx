import Notify from '../../enums/Notify';
import { IPreliminaryState } from './Types/IPreliminaryState';
import EventType from '../../enums/EventType';
export const initialState: IPreliminaryState = {
    preliminaryDetails: [],
    notify: Notify.none,
    event: EventType.none
  };