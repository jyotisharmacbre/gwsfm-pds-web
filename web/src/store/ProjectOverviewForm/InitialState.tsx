import { IProjectOverviewState } from './Types/IProjectOverviewState';
import Notify from '../../enums/Notify';
import EventType from '../../enums/EventType';

export const initialState: IProjectOverviewState = {
  form: {
    projectId: '',
    projectAdditionalDetail: {
      projectAddDetailId: '',
      projectId: '',
      mainContractor: '',
      enquiryReceivedFrom: '',
      enquiryTypeId: -1,
      creditCheckResult: '',
      siteAddress: '',
      cdmNotifiable: false,
      formOfContract: '',
      retention: '',
      liquidatedDamages: '',
      insurance: '',
      workTypeId: -1,
      commenceDate: new Date().toJSON(),
      completionDate: new Date().toJSON(),
      milestones: '',
      firstValuationDate: new Date().toJSON(),
      finalAccountDate: new Date().toJSON(),
      valuationIntervals: '',
      paymentTerms: '',
      isProjectLive: false,
      comments: '',
      authorizedByHop: '',
      budget: 1,
      authorizedBy: '',
      authorizedBySecond: '',
      authorizedByThird: '',
      projectRisk1: '',
      projectRisk2: '',
      projectRisk3: '',
      projectRiskControlMeasure1: '',
      projectRiskControlMeasure2: '',
      projectRiskControlMeasure3: ''
    },
    projectApprovals: []
  },
  error: null,
  loading: false,
  notify: Notify.none,
  event: EventType.none
};