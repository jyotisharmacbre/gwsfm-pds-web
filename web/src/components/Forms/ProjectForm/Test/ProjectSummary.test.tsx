import React from 'react';
import { shallow } from 'enzyme';
import ProjectSummary from '../ProjectSummary';
import { initialState as projectInitialState } from '../../../../store/CustomerEnquiryForm/InitialState';
import { findByTestAtrr } from '../../../../helpers/test-helper';

let wrapper: any;
const mountComponent = (props) => {
	wrapper = shallow(<ProjectSummary {...props} />);
};

describe('Project Summary componenet', () => {
	const props: any = {
		project: projectInitialState.form,
		lookUpData: [],
		company: 'test',
		headOfProject: 'test',
		projectOwner: 'test',
		projectManager: 'test'
	};

	beforeEach(() => {
		mountComponent(props);
	});

	it('should defines the componenet', () => {
		expect(wrapper).toBeDefined();
	});
	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should render the company', () => {
		expect(findByTestAtrr(wrapper, 'company')).toBeDefined();
	});

    it('should render the head of project', () => {
		expect(findByTestAtrr(wrapper, 'head-of-project')).toBeDefined();
	});

});
