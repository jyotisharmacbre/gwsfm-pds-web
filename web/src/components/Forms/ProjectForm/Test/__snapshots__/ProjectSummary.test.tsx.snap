// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Summary componenet should match the snapshot 1`] = `ShallowWrapper {}`;
